backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+import fk.prof.backend.util.ProtoUtil;
+import fk.prof.backend.util.ZookeeperUtil;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.curator.framework.CuratorFramework;
+import recording.Recorder;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class ZookeeperBasedBackendAssociationStore implements BackendAssociationStore {
+  private static Logger logger = LoggerFactory.getLogger(ZookeeperBasedBackendAssociationStore.class);
+
+  private final Vertx vertx;
+  private final CuratorFramework curatorClient;
+  private final String backendAssociationPath;
+  private final int reportingFrequencyInSeconds;
+  private final int maxAllowedSkips;
+
+  private final Map<String, BackendDetail> backendDetailLookup = new ConcurrentHashMap<>();
+  private final Map<Recorder.ProcessGroup, String> processGroupToBackendLookup = new ConcurrentHashMap<>();
+  private final PriorityQueue<BackendDetail> availableBackendsByPriority;
+  private final ReentrantLock backendAssignmentLock = new ReentrantLock();
+
+  private ZookeeperBasedBackendAssociationStore(Vertx vertx, CuratorFramework curatorClient, String backendAssociationPath,
+                                                int reportingFrequencyInSeconds, int maxAllowedSkips,
+                                                Comparator<BackendDetail> backendPriorityComparator)
+      throws Exception {
+    this.vertx = vertx;
+    this.curatorClient = curatorClient;
+    this.backendAssociationPath = backendAssociationPath;
+    this.reportingFrequencyInSeconds = reportingFrequencyInSeconds;
+    this.maxAllowedSkips = maxAllowedSkips;
+
+    List<BackendDetail> existingBackendsInZookeeper = loadDataFromZookeeper();
+    this.availableBackendsByPriority = new PriorityQueue<>(backendPriorityComparator);
+    this.availableBackendsByPriority.addAll(existingBackendsInZookeeper);
+
+    for(BackendDetail backendDetail: existingBackendsInZookeeper) {
+      this.backendDetailLookup.put(backendDetail.getBackendIPAddress(), backendDetail);
+      for(Recorder.ProcessGroup processGroup: backendDetail.getAssociatedProcessGroups()) {
+        if (this.processGroupToBackendLookup.putIfAbsent(processGroup, backendDetail.getBackendIPAddress()) != null) {
+          throw new IllegalStateException(String.format("Backend mapping already exists for process group=%s", processGroup));
+        }
+      }
+    }
+  }
+
+  @Override
+  public Future<Recorder.ProcessGroups> reportBackendLoad(BackendDTO.LoadReportRequest payload) {
+    String backendIPAddress = payload.getIp();
+    Future<Recorder.ProcessGroups> result = Future.future();
+    vertx.executeBlocking(future -> {
+      try {
+        BackendDetail backendDetail = backendDetailLookup.computeIfAbsent(backendIPAddress, (key) -> {
```

Comment:
there could be a separate cleanup job to delete defunct backends which cross some threshold?